Export SSO types and functions in lib/web

camscale opened the pull request:
Export some names in the `lib/web` package so that SSO auth plugins can be implemented from outside this package:

  * struct `ssoRequestParams` (including fields)
  * struct `ssoCallbackResponse` (including fields)
  * func `parseSSORequestParams`
  * func `ssoSetWebSessionAndRedirectURL`
  * func `redirectURLWithError`
  * var `ssoLoginConsoleErr`
  * type `CachedSessionLingeringThreshold` (for TestSAML)

Issue: https://github.com/gravitational/teleport.e/issues/525

---

Note: This supersedes PR #17457, exporting much less while still enabling the SAML lib/web tests to
move to the enterprise repo.

Follow-on PRs are:
* https://github.com/gravitational/teleport.e/pull/562 (Add SAML/OIDC web handlers to enterprise repo)
* https://github.com/gravitational/teleport/pull/17459 (Removes SAML/OIDC web handlers from OSS repo)

camscale committed:
Export SSO types and functions in lib/auth

Export some names in the `lib/auth` package so that SSO auth plugins can
be implemented from outside this package:

  * struct `ssoRequestParams` (including fields)
  * struct `ssoCallbackResponse` (including fields)
  * func `parseSSORequestParams`
  * func `ssoSetWebSessionAndRedirectURL`
  * func `redirectURLWithError`
  * var `ssoLoginConsoleErr`
  * type `CachedSessionLingeringThreshold` (for TestSAML)

camscale commented:
This PR supersedes https://github.com/gravitational/teleport/pull/17457, taking a cut-down approach to what is exported.

Tener commented on lib/web/apiserver.go:
```diff
@@ -1593,7 +1593,7 @@ func ConstructSSHResponse(response AuthParams) (*url.URL, error) {
 }
 
-func redirectURLWithError(clientRedirectURL string, errReply error) (*url.URL, error) {
+func RedirectURLWithError(clientRedirectURL string, errReply error) (*url.URL, error) {
```
Missing doc.

camscale replied:
done

Tener commented on lib/web/apiserver.go:
```diff
@@ -3167,13 +3167,13 @@ func makeTeleportClientConfig(ctx context.Context, sesCtx *SessionContext) (*cli
-	clientRedirectURL string
-	connectorID       string
-	csrfToken         string
+type SSORequestParams struct {
```
Missing doc for type and fields.

camscale replied:
done

Tener commented on lib/web/apiserver.go:
```diff
@@ -3167,13 +3167,13 @@ func makeTeleportClientConfig(ctx context.Context, sesCtx *SessionContext) (*cli
 }
 
-func parseSSORequestParams(r *http.Request) (*ssoRequestParams, error) {
+func ParseSSORequestParams(r *http.Request) (*SSORequestParams, error) {
```
Missing doc.

Tener commented on lib/web/apiserver.go:
```diff
@@ -3205,37 +3205,37 @@ func parseSSORequestParams(r *http.Request) (*ssoRequestParams, error) {
-	username          string
-	sessionName       string
-	clientRedirectURL string
+type SSOCallbackResponse struct {
```
Missing doc for type and fields.

camscale replied:
done

Tener commented on lib/web/apiserver.go:
```diff
@@ -3205,37 +3205,37 @@ func parseSSORequestParams(r *http.Request) (*ssoRequestParams, error) {
 }
 
-func ssoSetWebSessionAndRedirectURL(w http.ResponseWriter, r *http.Request, response *ssoCallbackResponse, verifyCSRF bool) error {
+func SSOSetWebSessionAndRedirectURL(w http.ResponseWriter, r *http.Request, response *SSOCallbackResponse, verifyCSRF bool) error {
```
Missing doc.

camscale replied:
done

Tener reviewed (approved):
Looks good, the only issue I can see is some missing docs.

camscale committed:
Add doc comments to newly-exported functions, types and fields.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -77,8 +77,8 @@ import (
 const (
-	// ssoLoginConsoleErr is a generic error message to hide revealing sso login failure msgs.
-	ssoLoginConsoleErr = "Failed to login. Please check Teleport's log for more details."
+	// SSOLoginConsoleErr is a generic error message to hide revealing sso login failure msgs.
```
```suggestion
	// SSOLoginConsoleErr is a generic error message avoid disclosing sensitive SSO failure messages.
```

When something is exported, it's extra important to get the name and documentation right.

Additionally, I would generally expect that something named `FooErr` implements `error`, and this is just a string. Maybe `SSOLoginFailureMessage` or something along those lines would be a better name.

camscale replied:
ack. renamed and updated doc string as suggested.

camscale committed:
Rename SSOLoginConsoleErr to SSOLoginFailureMessage

Rename SSOLoginConsoleErr to SSOLoginFailureMessage and update the doc
string to be grammatically correct.

Addresses: https://github.com/gravitational/teleport/pull/17530#discussion_r998311851

camscale committed:
Merge branch 'master' into camh/export-sso

camscale committed:
Merge branch 'master' into camh/export-sso
